projects/extension/sql/idempotent/001-openai.sql

```diff
@@ -19,196 +19,391 @@ set search_path to pg_catalog, pg_temp
+    if _extra_body is not None:
+        kwargs['extra_body'] = json.loads(_extra_body)
+
+    async def async_openai_call(client, kwargs):
+        response = await client.models.with_raw_response.list(**kwargs)
+        return response.text
+
+    # Execute the API call with cancellation support
+    result = ai.openai.execute_with_cancellation(plpy, client, async_openai_call, **kwargs)
+
+    return result
 $python$
-language plpython3u volatile parallel safe security invoker
-set search_path to pg_catalog, pg_temp
+    language plpython3u volatile parallel unsafe security invoker
+                        set search_path to pg_catalog, pg_temp
 ;
 
 -------------------------------------------------------------------------------
 -- openai_embed
 -- generate an embedding from a text value
 -- https://platform.openai.com/docs/api-reference/embeddings/create
 create or replace function ai.openai_embed
-( model text
-, input_text text
-, api_key text default null
-, api_key_name text default null
-, base_url text default null
-, dimensions int default null
-, openai_user text default null
-) returns @extschema:vector@.vector
+( _input text
+, _model text
+, _api_key text DEFAULT NULL
+, _api_key_name text DEFAULT NULL
+, _base_url text DEFAULT NULL
+, _encoding_format text DEFAULT NULL
+, _dimensions int DEFAULT NULL
+, _user text DEFAULT NULL
+, _extra_headers jsonb DEFAULT NULL
+, _extra_query jsonb DEFAULT NULL
+, _extra_body jsonb DEFAULT NULL
+, _timeout float8 DEFAULT NULL
+) returns jsonb
 as $python$
     #ADD-PYTHON-LIB-DIR
     import ai.openai
-    for tup in ai.openai.embed(plpy, model, input_text, api_key=api_key, api_key_name=api_key_name, base_url=base_url, dimensions=dimensions, user=openai_user):
-        return tup[1]
+    import json
+
+    # Create async client
+    client = ai.openai.get_or_create_client(plpy, GD, _api_key, _api_key_name, _base_url)
+
+    # Prepare kwargs for the API call
+    kwargs = ai.openai.prepare_kwargs({
+        "input": [_input],
+        "model": _model,
+        "encoding_format": _encoding_format,
+        "dimensions": _dimensions,
+        "user": _user,
+    })
+
+    # Add extra parameters if provided
+    if _extra_headers is not None:
+        kwargs['extra_headers'] = json.loads(_extra_headers)
+    if _extra_query is not None:
+        kwargs['extra_query'] = json.loads(_extra_query)
+    if _extra_body is not None:
+        kwargs['extra_body'] = json.loads(_extra_body)
+
+    async def async_openai_call(client, kwargs):
+        response = await client.embeddings.with_raw_response.create(**kwargs)
+        return response.text
+
+    # Execute the API call with cancellation support
+    result = ai.openai.execute_with_cancellation(plpy, client, async_openai_call, **kwargs)
+
+    return result
 $python$
-language plpython3u immutable parallel safe security invoker
-set search_path to pg_catalog, pg_temp
+    language plpython3u volatile parallel unsafe security invoker
```

Comment:
why is this now volatile? That seems wrong. A call with the same parameters should always return the same embedding.

Reply:
So, the embedding endpoint for OpenAI is not totally deterministic. You can get slightly different vectors back for the same input on the same endpoint with the same model.

I'd say that is more of a bug with their implementation though, and immutable would work and be _much_ cheaper in a lot of situations (with duplicated data).
Will change it back.